src/server/routes/welcome.ts

```diff
@@ -273,6 +288,203 @@ router.get(
+					...consentsData,
+				},
+			} as RequestState);
+
+			trackMetric('NewAccountReview::Success');
+		} catch (error) {
+			logger.error(`${req.method} ${req.originalUrl}  Error`, error, {
+				request_id: res.locals.requestId,
+			});
+
+			const { message } = error instanceof ApiError ? error : new ApiError();
+
+			state = mergeRequestState(state, {
+				globalMessage: {
+					error: message,
+				},
+			});
+
+			trackMetric('NewAccountReview::Failure');
+		}
+
+		const html = renderer('/welcome/review', {
+			pageTitle: 'Your data',
+			requestState: state,
+		});
+		return res.type('html').send(html);
+	}),
+);
+
+router.get(
+	'/welcome/newsletters',
+	loginMiddlewareOAuth,
+	async (req: Request, res: ResponseWithRequestState) => {
+		const state = res.locals;
+		const geolocation = state.pageData.geolocation;
+		const newsletters = await getUserNewsletterSubscriptions({
+			newslettersOnPage: NewsletterMap.get(geolocation) as string[],
+			ip: req.ip,
+			sc_gu_u: req.cookies.SC_GU_U,
+			request_id: state.requestId,
+			accessToken: state.oauthState.accessToken?.toString(),
+		});
+		// eslint-disable-next-line functional/immutable-data
+		state.pageData.newsletters = newsletters;
+		const html = renderer('/welcome/newsletters', {
+			pageTitle: 'Choose Newsletters',
+			requestState: state,
+		});
+		res.type('html').send(html);
+	},
+);
+
+router.post(
+	'/welcome/review',
+	loginMiddlewareOAuth,
+	handleAsyncErrors(async (req: Request, res: ResponseWithRequestState) => {
+		const state = res.locals;
+		const { returnUrl, fromURI } = state.queryParams;
+
+		const sc_gu_u = req.cookies.SC_GU_U;
+
+		try {
+			// Attempt to update location for consented users.
+			if (res.locals.oauthState.accessToken) {
+				await updateRegistrationLocationViaOkta(
+					req,
+					res.locals.oauthState.accessToken,
+				);
+			} else {
+				await updateRegistrationLocationViaIDAPI(req.ip, sc_gu_u, req);
+			}
+
+			const consents = CONSENTS_DATA_PAGE.map((id) => ({
+				id,
+				consented: getConsentValueFromRequestBody(id, req.body),
+			}));
+
+			await patchConsents({
+				ip: req.ip,
+				sc_gu_u,
+				accessToken: res.locals.oauthState.accessToken?.toString(),
+				payload: consents,
+				request_id: res.locals.requestId,
+			});
+
+			trackMetric('NewAccountReviewSubmit::Success');
+		} catch (error) {
+			// Never block the user at this point, so we'll just log the error
+			logger.error(`${req.method} ${req.originalUrl}  Error`, error, {
+				request_id: res.locals.requestId,
+			});
+
+			trackMetric('NewAccountReviewSubmit::Failure');
+		} finally {
+			// if there is a fromURI, we need to complete the oauth flow, so redirect to the fromURI
+			if (fromURI) {
+				return res.redirect(303, fromURI);
+			}
+
+			// In all other cases, select the next page to redirect to based on geolocation
+			const nextPage = getNextWelcomeFlowPage({
+				geolocation: state.pageData.geolocation,
+				returnUrl,
+				queryParams: state.queryParams,
+			});
```

Comment:
hmm.. glad you remembered the `fromURI` here!

But don't we want to show the newsletters page for US and AU even if they have a `fromURI`?

it could be reworked where (in pseudocode)

IF `fromURI` AND NOT (geolocation US OR geolocation AU)
  THEN return `fromURI`
ELSE
	CONTINUE

and then in the `POST /welcome/newsletters` handle the `fromURI` too in the finally block

this way we can still get users getting newsletters even on apps and not just web, we're already showing the review page anyway, so can also add the newsletter page surely for these users!

Reply:
We've decided for the moment to leave the `fromURI` logic unchanged, so for apps, the newsletters page never gets shown. I have however refactored the `fromURI` check into `getNextWelcomeFlowPage`, to make it easier to keep track of where all the logic happens.